stream/replicator.h

```diff
@@ -80,36 +82,155 @@ class SubscribableRemoteStream final {
+          subscriber_(subscriber),
+          index_(start_idx),
+          from_us_(from_us),
+          unused_idxts_() {}
+
+    void PassChunkToSubscriber(const std::string& chunk) {
+      const size_t chunk_size = chunk.size();
+      size_t end_pos = 0;
+      if (!carried_over_data_.empty()) {
+        while (end_pos < chunk_size && chunk[end_pos] != '\n' && chunk[end_pos] != '\r') {
+          ++end_pos;
+        }
+        if (end_pos == chunk_size) {
+          carried_over_data_ += chunk;
+          return;
+        }
+        PassEntryToSubscriber(carried_over_data_ + chunk.substr(0, end_pos));
+      }
+
+      size_t start_pos = end_pos;
+      for (;;) {
+        while (start_pos < chunk_size && (chunk[start_pos] == '\n' || chunk[start_pos] == '\r')) {
+          ++start_pos;
+        }
+        end_pos = start_pos + 1;
+        while (end_pos < chunk_size && chunk[end_pos] != '\n' && chunk[end_pos] != '\r') {
+          ++end_pos;
+        }
+        if (end_pos >= chunk_size) {
+          break;
+        }
+        PassEntryToSubscriber(chunk.substr(start_pos, end_pos - start_pos));
+        start_pos = end_pos + 1;
+      }
+      if (start_pos < chunk_size) {
+        carried_over_data_ = chunk.substr(start_pos);
+      } else {
+        carried_over_data_.clear();
+      }
+    }
+
+   protected:
+    BorrowedWithCallback<RemoteStream> borrowed_remote_stream_;
+    F& subscriber_;
+    uint64_t index_;
+    std::chrono::microseconds from_us_;
+    const idxts_t unused_idxts_;
+    std::string carried_over_data_;
+
+   private:
+    template <ReplicationMode MODE = RM>
+    ENABLE_IF<MODE == ReplicationMode::Checked> PassEntryToSubscriber(const std::string& raw_log_line) {
+      const auto split = current::strings::Split(raw_log_line, '\t');
+      if (split.empty()) {
+        CURRENT_THROW(RemoteStreamMalformedChunkException());
+      }
+      const auto tsoptidx = ParseJSON<ts_optidx_t>(split[0]);
+      if (Exists(tsoptidx.index)) {
+        const auto idxts = idxts_t(Value(tsoptidx.index), tsoptidx.us);
+        if (split.size() != 2u || idxts.index != index_ || (from_us_.count() > 0 && idxts.us < from_us_)) {
+          CURRENT_THROW(RemoteStreamMalformedChunkException());
+        }
+        auto entry = ParseJSON<TYPE_SUBSCRIBED_TO>(split[1]);
+        ++index_;
+        from_us_ = std::chrono::microseconds(0);
+        if (subscriber_(std::move(entry), idxts, unused_idxts_) == ss::EntryResponse::Done) {
+          CURRENT_THROW(StreamTerminatedBySubscriber());
+        }
+      } else {
+        if (split.size() != 1u || (from_us_.count() > 0 && tsoptidx.us < from_us_)) {
+          CURRENT_THROW(RemoteStreamMalformedChunkException());
+        }
+        if (subscriber_(tsoptidx.us) == ss::EntryResponse::Done) {
+          CURRENT_THROW(StreamTerminatedBySubscriber());
+        }
+        from_us_ = tsoptidx.us + std::chrono::microseconds(1);
```

Comment:
Do we allow multiple `UpdateHead`-s with the same timestamp (and will they be persisted)? If yes, this may fail.